math/minuit2/inc/Minuit2/NumericalDerivator.h

```diff
@@ -0,0 +1,119 @@
+// @(#)root/mathcore:$Id$
+// Authors: L. Moneta, J.T. Offermann, E.G.P. Bos    2013-2018
+//
+/**********************************************************************
+ *                                                                    *
+ * Copyright (c) 2013 , LCG ROOT MathLib Team                         *
+ *                                                                    *
+ **********************************************************************/
+/*
+ * NumericalDerivator.h
+ *
+ *  Original version created on: Aug 14, 2013
+ *      Authors: L. Moneta, J. T. Offermann
+ *  Modified version created on: Sep 27, 2017
+ *      Author: E. G. P. Bos
+ */
+
+#ifndef ROOT_Minuit2_NumericalDerivator
+#define ROOT_Minuit2_NumericalDerivator
+
+#ifndef ROOT_Math_IFunctionfwd
+#include <Math/IFunctionfwd.h>
+#endif
+
+#include <vector>
+#include "Fit/ParameterSettings.h"
+#include "Minuit2/SinParameterTransformation.h"
+#include "Minuit2/SqrtUpParameterTransformation.h"
+#include "Minuit2/SqrtLowParameterTransformation.h"
+#include "Minuit2/MnMachinePrecision.h"
+
+namespace ROOT {
+namespace Minuit2 {
+
+// Holds all necessary derivatives and associated numbers (per parameter) used in the NumericalDerivator class.
+struct DerivatorElement {
+   double derivative;
+   double second_derivative;
+   double step_size;
+};
+
+class NumericalDerivator {
+public:
+   explicit NumericalDerivator(bool always_exactly_mimic_minuit2 = true);
+   NumericalDerivator(const NumericalDerivator &other);
+   NumericalDerivator(double step_tolerance, double grad_tolerance, unsigned int ncycles, double error_level,
+                             bool always_exactly_mimic_minuit2 = true);
+   virtual ~NumericalDerivator();
+
+   void setup_differentiate(const ROOT::Math::IBaseFunctionMultiDim *function, const double *cx,
```

Comment:
Can you rename to `SetupDifferentiate` to be consistent with ROOT naming conventions ?